R/for_loop_index_linter.R

```diff
@@ -34,10 +34,7 @@ for_loop_index_linter <- function() {
   //forcond
     /SYMBOL[text() =
       following-sibling::expr
-        //SYMBOL[not(
-          preceding-sibling::OP-DOLLAR
-          or parent::expr[preceding-sibling::OP-LEFT-BRACKET]
-        )]
+        //SYMBOL[not(parent::expr[OP-DOLLAR or OP-AT or preceding-sibling::OP-LEFT-BRACKET])]
```

Comment:
The `//` seems off to me. Shouldn't this be a single slash?

Reply:
I think the intent is "don't overwrite any symbol in the parent env":

https://github.com/r-lib/lintr/blob/bb7669afa4b90843e8358397e60d02e2fc512953/tests/testthat/test-for_loop_index_linter.R#L19

Maybe this isn't the best approach but it seems to work OK. Of course there'll be false positives from NSE column names... Let's leave as is until an issue is filed?

Reply:
I'm expecting this to trigger a false positive for the expression

```r
foo <- function() { # force single expression
  for (i in 1:10) {
    42L
  }
  i <- 7L
}
```

I think instead using `/descendant::expr` will restrict (correctly) to the RHS of the forcond.

Not sure though, I'm on my phone.
We can leave this in and file a separate issue if my suspicion is correct since this PR has a different scope and the old code had a // as well.

Reply:
no lint for that, just checked. i'll add it as a regression test. I think `//` is the same as `.//` inside `[`. I will use `.//` to make it clearer.

Reply:
oh wait, read more closely. Note that it's

```xpath
following-sibling::expr
  //SYMBOL
```

i.e. `following-sibling::expr//SYMBOL`